src/models/kv_cache.h

```diff
@@ -4,8 +4,18 @@
 
 namespace Generators {
 
-struct KV_Cache_Combined {
-  KV_Cache_Combined(State& state);
+struct KeyValueCache {
+  virtual ~KeyValueCache() = default;
+  virtual void Add() = 0;
+  virtual void AddEncoder() = 0;
+  virtual void Update(DeviceSpan<int32_t> beam_indices, int total_length) = 0;
+  virtual void RewindTo(size_t index) = 0;
+
+  static bool IsCacheNeeded(const Model& model);
+};
+
+struct KeyValueCacheDefault_Combined {
```

Comment:
Their usage is limited to files that are not touched by this PR and so I didn't update them. But I'll do that now since it might get ugly to do this later.

Reply:
Adding to the above:

- KeyValueCacheDefault_Combined is only used in the gpt model type. This type of key-value cache should be deprecated going forward. Not sure when we will remove the code references, but we should not be using that. But, I will make the changes to that struct to inherit from KeyValuCache
- Cross_Cache is only used in whisper. This type of key-value cache is not used as a standalone cache and is a supplemental cache for the whisper model needed in some scenarios. It does not behave like the other KeyValue caches. I don't think I will update this one since it might have some implications on how whisper model uses it.